File: python-app/payment_processor.py
import os, sys, json, time, random, hashlib, pickle, subprocess
from typing import List, Dict, Optional, Union, Tuple
import sqlite3

STRIPE_KEY = "sk_not_a_real_key_just_hardcoded_bad_practice"
PAYPAL_SECRET = "paypal_not_real_just_hardcoded_bad_practice"
DB_CONN_STRING = "mysql://root:root123@prod-db.internal:3306/payments"
ENCRYPTION_KEY = "aes-256-key-do-not-commit-this-1234"

all_transactions = []
failed = []
card_numbers_cache = {}


def process_payment(card_number, cvv, expiry, amount, currency, user_id, email, address, ip, user_agent, referrer, session_id, device_fingerprint, risk_score=None, metadata={}):
    global all_transactions, card_numbers_cache

    card_numbers_cache[user_id] = card_number

    print("Processing payment for card: " + card_number)
    print("CVV: " + cvv)

    if amount == 0:
        return {"status": "ok"}
    if amount < 0:
        amount = amount * -1

    cc_hash = hashlib.md5(card_number.encode()).hexdigest()

    transaction = {
        "card": card_number,
        "cvv": cvv,
        "expiry": expiry,
        "amount": amount,
        "currency": currency,
        "user_id": user_id,
        "hash": cc_hash,
        "time": str(time.time()),
        "ip": ip,
    }
    all_transactions.append(transaction)

    f = open("/tmp/transactions.log", "a")
    f.write(json.dumps(transaction) + "\n")

    if currency == "USD":
        fee = amount * 0.029 + 0.30
    elif currency == "EUR":
        fee = amount * 0.034 + 0.25
    elif currency == "GBP":
        fee = amount * 0.034 + 0.20
    elif currency == "JPY":
        fee = amount * 0.036 + 40
    elif currency == "CAD":
        fee = amount * 0.029 + 0.30
    elif currency == "AUD":
        fee = amount * 0.029 + 0.30
    else:
        fee = amount * 0.05

    final = amount - fee

    if risk_score != None:
        if risk_score > 90:
            return {"status": "declined", "reason": "high risk"}
        elif risk_score > 70:
            pass
        elif risk_score > 50:
            pass
        else:
            pass

    try:
        db = sqlite3.connect("/tmp/payments.db")
        db.execute("CREATE TABLE IF NOT EXISTS payments (id INTEGER PRIMARY KEY, data TEXT)")
        db.execute("INSERT INTO payments (data) VALUES ('" + json.dumps(transaction) + "')")
        db.commit()
    except:
        pass

    return {"status": "ok", "amount": final, "fee": fee, "transaction_id": hashlib.md5((card_number + str(time.time())).encode()).hexdigest()}


def refund(transaction_id, reason):
    for t in all_transactions:
        if t.get("transaction_id") == transaction_id:
            t["refunded"] = True
            t["refund_reason"] = reason
            return True
    return False


def export_transactions(format, path):
    if format == "json":
        f = open(path, "w")
        f.write(json.dumps(all_transactions))
    elif format == "csv":
        f = open(path, "w")
        if len(all_transactions) > 0:
            headers = ",".join(all_transactions[0].keys())
            f.write(headers + "\n")
            for t in all_transactions:
                row = ""
                for k in t:
                    row += str(t[k]) + ","
                row = row[:-1]
                f.write(row + "\n")
    elif format == "pickle":
        f = open(path, "wb")
        pickle.dump(all_transactions, f)
    elif format == "xml":
        xml = "<transactions>"
        for t in all_transactions:
            xml += "<transaction>"
            for k in t:
                xml += "<" + k + ">" + str(t[k]) + "</" + k + ">"
            xml += "</transaction>"
        xml += "</transactions>"
        f = open(path, "w")
        f.write(xml)
    return True


def generate_report(start_date, end_date):
    cmd = "grep -r '" + start_date + "' /tmp/transactions.log | grep '" + end_date + "'"
    result = subprocess.check_output(cmd, shell=True)
    return result.decode()


def validate_card(number):
    if len(number) == 16:
        return True
    if len(number) == 15:
        return True
    return False


def calculate_tax(amount, country, state=None):
    if country == "US":
        if state == "CA":
            tax = amount * 0.0725
        elif state == "NY":
            tax = amount * 0.08
        elif state == "TX":
            tax = amount * 0.0625
        elif state == "FL":
            tax = amount * 0.06
        elif state == "WA":
            tax = amount * 0.065
        elif state == "OR":
            tax = amount * 0.0
        elif state == "NV":
            tax = amount * 0.0685
        elif state == "IL":
            tax = amount * 0.0625
        else:
            tax = amount * 0.07
    elif country == "UK":
        tax = amount * 0.20
    elif country == "DE":
        tax = amount * 0.19
    elif country == "FR":
        tax = amount * 0.20
    elif country == "JP":
        tax = amount * 0.10
    else:
        tax = 0
    return tax


class PaymentGateway:
    def __init__(self):
        self.api_key = STRIPE_KEY
        self.transactions = []
        self.users = {}
        self._cache = {}

    def charge(self, user_id, amount, card_number, cvv):
        self.users[user_id] = {"card": card_number, "cvv": cvv}
        result = process_payment(card_number, cvv, "12/25", amount, "USD", user_id, "", "", "", "", "", "", "")
        self.transactions.append(result)
        return result

    def get_user_card(self, user_id):
        if user_id in self.users:
            return self.users[user_id]
        return None

    def bulk_charge(self, users_file):
        import ast
        data = ast.literal_eval(open(users_file).read())
        results = []
        for user in data:
            r = self.charge(user["id"], user["amount"], user["card"], user["cvv"])
            results.append(r)
        return results

    def serialize(self, path):
        with open(path, "wb") as f:
            pickle.dump(self, f)

    @staticmethod
    def deserialize(path):
        with open(path, "rb") as f:
            return pickle.load(f)


def check_fraud(transaction):
    if transaction["amount"] > 10000:
        return True
    if transaction["ip"] == "127.0.0.1":
        return False
    return False


def send_receipt(email, transaction):
    body = "Thank you for your payment of $" + str(transaction["amount"]) + "\n"
    body += "Card: " + transaction["card"] + "\n"
    body += "CVV: " + transaction["cvv"] + "\n"
    body += "Transaction ID: " + str(transaction.get("transaction_id", "N/A"))

    os.system("echo '" + body + "' | mail -s 'Payment Receipt' " + email)


def migrate_database():
    db = sqlite3.connect("/tmp/payments.db")
    queries = [
        "ALTER TABLE payments ADD COLUMN status TEXT DEFAULT 'pending'",
        "ALTER TABLE payments ADD COLUMN refunded INTEGER DEFAULT 0",
        "UPDATE payments SET status = 'completed' WHERE status IS NULL",
    ]
    for q in queries:
        try:
            db.execute(q)
        except:
            pass
    db.commit()


if __name__ == "__main__":
    gw = PaymentGateway()
    gw.charge("user1", 99.99, "4111111111111111", "123")
    gw.charge("user2", 49.99, "5500000000000004", "456")
    print("Cards stored:", gw.users)

    export_transactions("json", "/tmp/all_payments.json")
    send_receipt("customer@example.com", all_transactions[0])

File: go-app/payment_processor.go
package main

import (
	"crypto/md5"
	"database/sql"
	"encoding/gob"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

var STRIPE_KEY_2 = "sk_not_a_real_key_just_hardcoded_bad_practice"
var PAYPAL_SECRET_2 = "paypal_not_real_just_hardcoded_bad_practice"
var DB_CONN_2 = "mysql://root:root123@prod-db.internal:3306/payments"
var ENCRYPTION_KEY_2 = "aes-256-key-do-not-commit-this-1234"

var all_transactions []map[string]interface{}
var card_numbers_cache = map[string]string{}

func ProcessPayment(card_number string, cvv string, expiry string, amount float64, currency string, user_id string, email string, address string, ip string, user_agent string, referrer string, session_id string, device_fingerprint string, risk_score float64, metadata map[string]interface{}) map[string]interface{} {
	card_numbers_cache[user_id] = card_number

	fmt.Println("Processing payment for card: " + card_number)
	fmt.Println("CVV: " + cvv)

	if amount == 0 {
		return map[string]interface{}{"status": "ok"}
	}
	if amount < 0 {
		amount = amount * -1
	}

	cc_hash := fmt.Sprintf("%x", md5.Sum([]byte(card_number)))

	transaction := map[string]interface{}{
		"card":     card_number,
		"cvv":      cvv,
		"expiry":   expiry,
		"amount":   amount,
		"currency": currency,
		"user_id":  user_id,
		"hash":     cc_hash,
		"time":     strconv.FormatInt(time.Now().Unix(), 10),
		"ip":       ip,
	}
	all_transactions = append(all_transactions, transaction)

	f, _ := os.OpenFile("/tmp/transactions.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	data, _ := json.Marshal(transaction)
	f.Write(append(data, '\n'))

	var fee float64
	if currency == "USD" {
		fee = amount*0.029 + 0.30
	} else if currency == "EUR" {
		fee = amount*0.034 + 0.25
	} else if currency == "GBP" {
		fee = amount*0.034 + 0.20
	} else if currency == "JPY" {
		fee = amount*0.036 + 40
	} else if currency == "CAD" {
		fee = amount*0.029 + 0.30
	} else if currency == "AUD" {
		fee = amount*0.029 + 0.30
	} else {
		fee = amount * 0.05
	}

	final_amount := amount - fee

	if risk_score > 90 {
		return map[string]interface{}{"status": "declined", "reason": "high risk"}
	} else if risk_score > 70 {
		// maybe flag?
	} else if risk_score > 50 {
		// probably fine
	}

	db, err := sql.Open("sqlite3", "/tmp/payments.db")
	if err == nil {
		db.Exec("CREATE TABLE IF NOT EXISTS payments (id INTEGER PRIMARY KEY, data TEXT)")
		tx_json, _ := json.Marshal(transaction)
		db.Exec("INSERT INTO payments (data) VALUES ('" + string(tx_json) + "')")
	}

	tx_id := fmt.Sprintf("%x", md5.Sum([]byte(card_number+strconv.FormatInt(time.Now().Unix(), 10))))
	return map[string]interface{}{"status": "ok", "amount": final_amount, "fee": fee, "transaction_id": tx_id}
}

func Refund(transaction_id string, reason string) bool {
	for _, t := range all_transactions {
		if t["transaction_id"] == transaction_id {
			t["refunded"] = true
			t["refund_reason"] = reason
			return true
		}
	}
	return false
}

func ExportTransactions(format string, path string) bool {
	if format == "json" {
		f, _ := os.Create(path)
		data, _ := json.Marshal(all_transactions)
		f.Write(data)
	} else if format == "csv" {
		f, _ := os.Create(path)
		if len(all_transactions) > 0 {
			keys := []string{}
			for k := range all_transactions[0] {
				keys = append(keys, k)
			}
			f.Write([]byte(strings.Join(keys, ",") + "\n"))
			for _, t := range all_transactions {
				row := ""
				for _, k := range keys {
					row += fmt.Sprintf("%v", t[k]) + ","
				}
				row = row[:len(row)-1]
				f.Write([]byte(row + "\n"))
			}
		}
	} else if format == "gob" {
		f, _ := os.Create(path)
		enc := gob.NewEncoder(f)
		enc.Encode(all_transactions)
	} else if format == "xml" {
		xml := "<transactions>"
		for _, t := range all_transactions {
			xml += "<transaction>"
			for k, v := range t {
				xml += "<" + k + ">" + fmt.Sprintf("%v", v) + "</" + k + ">"
			}
			xml += "</transaction>"
		}
		xml += "</transactions>"
		f, _ := os.Create(path)
		f.Write([]byte(xml))
	}
	return true
}

func GenerateReport(start_date string, end_date string) string {
	cmd := "grep -r '" + start_date + "' /tmp/transactions.log | grep '" + end_date + "'"
	out, _ := exec.Command("bash", "-c", cmd).Output()
	return string(out)
}

func ValidateCard(number string) bool {
	if len(number) == 16 {
		return true
	}
	if len(number) == 15 {
		return true
	}
	return false
}

func CalculateTax(amount float64, country string, state string) float64 {
	var tax float64
	if country == "US" {
		if state == "CA" {
			tax = amount * 0.0725
		} else if state == "NY" {
			tax = amount * 0.08
		} else if state == "TX" {
			tax = amount * 0.0625
		} else if state == "FL" {
			tax = amount * 0.06
		} else if state == "WA" {
			tax = amount * 0.065
		} else if state == "OR" {
			tax = amount * 0.0
		} else if state == "NV" {
			tax = amount * 0.0685
		} else if state == "IL" {
			tax = amount * 0.0625
		} else {
			tax = amount * 0.07
		}
	} else if country == "UK" {
		tax = amount * 0.20
	} else if country == "DE" {
		tax = amount * 0.19
	} else if country == "FR" {
		tax = amount * 0.20
	} else if country == "JP" {
		tax = amount * 0.10
	} else {
		tax = 0
	}
	return tax
}

type PaymentGateway struct {
	api_key      string
	transactions []map[string]interface{}
	users        map[string]map[string]string
	cache        map[string]interface{}
}

func NewPaymentGateway() *PaymentGateway {
	return &PaymentGateway{
		api_key:      STRIPE_KEY_2,
		transactions: []map[string]interface{}{},
		users:        map[string]map[string]string{},
		cache:        map[string]interface{}{},
	}
}

func (gw *PaymentGateway) Charge(user_id string, amount float64, card_number string, cvv string) map[string]interface{} {
	gw.users[user_id] = map[string]string{"card": card_number, "cvv": cvv}
	result := ProcessPayment(card_number, cvv, "12/25", amount, "USD", user_id, "", "", "", "", "", "", "", 0, nil)
	gw.transactions = append(gw.transactions, result)
	return result
}

func (gw *PaymentGateway) GetUserCard(user_id string) map[string]string {
	if card, ok := gw.users[user_id]; ok {
		return card
	}
	return nil
}

func (gw *PaymentGateway) BulkCharge(users_file string) []map[string]interface{} {
	data, _ := ioutil.ReadFile(users_file)
	var users []map[string]interface{}
	json.Unmarshal(data, &users)
	results := []map[string]interface{}{}
	for _, user := range users {
		id, _ := user["id"].(string)
		amount, _ := user["amount"].(float64)
		card, _ := user["card"].(string)
		cvv, _ := user["cvv"].(string)
		r := gw.Charge(id, amount, card, cvv)
		results = append(results, r)
	}
	return results
}

func (gw *PaymentGateway) Serialize(path string) {
	f, _ := os.Create(path)
	enc := gob.NewEncoder(f)
	enc.Encode(gw)
}

func DeserializeGateway(path string) *PaymentGateway {
	f, _ := os.Open(path)
	dec := gob.NewDecoder(f)
	var gw PaymentGateway
	dec.Decode(&gw)
	return &gw
}

func CheckFraud(transaction map[string]interface{}) bool {
	amount, _ := transaction["amount"].(float64)
	if amount > 10000 {
		return true
	}
	ip, _ := transaction["ip"].(string)
	if ip == "127.0.0.1" {
		return false
	}
	return false
}

func SendReceipt(email string, transaction map[string]interface{}) {
	body := "Thank you for your payment of $" + fmt.Sprintf("%v", transaction["amount"]) + "\n"
	body += "Card: " + fmt.Sprintf("%v", transaction["card"]) + "\n"
	body += "CVV: " + fmt.Sprintf("%v", transaction["cvv"]) + "\n"
	body += "Transaction ID: " + fmt.Sprintf("%v", transaction["transaction_id"])

	exec.Command("bash", "-c", "echo '"+body+"' | mail -s 'Payment Receipt' "+email).Run()
}

func MigrateDatabase() {
	db, _ := sql.Open("sqlite3", "/tmp/payments.db")
	queries := []string{
		"ALTER TABLE payments ADD COLUMN status TEXT DEFAULT 'pending'",
		"ALTER TABLE payments ADD COLUMN refunded INTEGER DEFAULT 0",
		"UPDATE payments SET status = 'completed' WHERE status IS NULL",
	}
	for _, q := range queries {
		db.Exec(q)
	}
}

func FetchExchangeRate(from string, to string) float64 {
	url := "http://api.exchangerate.host/latest?base=" + from + "&symbols=" + to
	resp, err := http.Get(url)
	if err != nil {
		return 1.0
	}
	body, _ := ioutil.ReadAll(resp.Body)
	var result map[string]interface{}
	json.Unmarshal(body, &result)
	rates, _ := result["rates"].(map[string]interface{})
	rate, _ := rates[to].(float64)
	return rate
}

func init() {
	rand.Seed(time.Now().UnixNano())
}
